grading_algorithm.py

```diff
@@ -0,0 +1,163 @@
+def grade_expression(expression, c_exp) :

```

Comment:
Type hint the function. I would also rename `expression` to `reformulation` and `c_exp` to `correct_expression`. I think it would make the code much more readable in the end.